posts/Swift-Optionals-Are-Monads-Too.md

```diff
@@ -0,0 +1,264 @@
+  case .Some(let x): return try f(x)
+  case .None: return .None
+}
+```
+
+Comparing flatMap and map on Optional you might be asking yourself what's the difference? Really there is only one specific difference and it's that f in map cannot return another Optional whereas f in flatMap can return an Optional.
+
+---
+
+# Why It Matters
+
+So now that we have a good understanding of what each of these abstract concepts mean, it kind of begs the question, what are they good for?
+
+### Mapping out Optionals
+
+Formatting NSDate? can be a bit tricky. Because it's an Optional we are forced into using if/let syntax or heaven forbid force unwrapping the Optional.
+
+```swift
+var date: NSDate?
+var formatted: NSString?
+
+// some time after date *may* have been set or initialized
+
+if let date = date {
+  formatted = NSDateFormatter().stringFromDate(date)
+}
+```
+
+Here's functionally equivalent code using `map`:
+
+```swift
+var date: NSDate?
+
+// some time after date *may* have been set or initialized
+
+var formatted = date.map(NSDateFormatter().stringFromDate)
+```
+
+To be clear, we aren't aiming for code golf. But switching over from the `if/let` to using `map` makes our intent more clear in my opinion. If the date isn't nil, let's format it as a string. If it is nil, our formatted date string is nil. 
+
+Another wonderful usage of `map` is while we are interpolating values in a string:
+
+```swift
+var numberOfFriends: Int?
+
+// some time after numberOfFriends *may* 
+// have been set or initialized
+
+var friendsStatus = numberOfFriends.map { 
+  "You have \($0) friends."
+} ?? "No friends yet." 
+```
+
+This allows us to choose two completely different strings based on whether our `numberOfFriends` variable has a value or not.
+
+One last example of a great usage for `map`. What if we want to read the contents of a url?
+
+```swift
+var webURL: NSURL? = NSURL(string: "http://rouxbe.com/")
+
+var webContents = webURL.map {
+  NSData(contentsOfURL: $0)
+}.map {
+  NSString(data: $0, encoding: NSUTF8StringEncoding)
+}
+
+webContents.map { print($0) }
+```
+
+If I was a bit more mischievous, I could leave the example like that and go on with the post. But the most astute reader who really gets these concepts would point out the mistake in the code sample above. In case you didn't catch it, that's ok. The problem with our code above is that map accepts a closure/function that does not return an `Optional`. `NSData`'s initializer that accepts an `NSURL` is a failable initializer. Since it can return `nil`, our closure can return `nil`. Xcode will complain inside our second call to `map` that `$0` has not been unwrapped. So what do we do? 
+
+### FlatMap to the Rescue
+```swift
+var webURL: NSURL? = NSURL(string: "http://rouxbe.com/")
+
+var webContents = webURL.flatMap {
+  NSData(contentsOfURL: $0)
+}.flatMap {
+  NSString(data: $0, encoding: NSUTF8StringEncoding)
+}
+
+webContents.map { print($0) }
+```
+
+### How do we really apply
```

Comment:
I was aiming for a little bit on the light side instead of sticking with something specific and to the point.